core/app/[locale]/(default)/product/[slug]/_actions/wishlist-action.ts

```diff
@@ -0,0 +1,304 @@
+import { SubmissionResult } from '@conform-to/react';
+import { parseWithZod } from '@conform-to/zod';
+import { revalidateTag } from 'next/cache';
+import { getLocale, getTranslations } from 'next-intl/server';
+import { cache } from 'react';
+import { z } from 'zod';
+
+import { CreateWishlistMutation } from '~/app/[locale]/(default)/account/wishlists/_actions/mutation';
+import { newWishlist } from '~/app/[locale]/(default)/account/wishlists/_actions/new-wishlist';
+import { getSessionCustomerAccessToken } from '~/auth';
+import { client } from '~/client';
+import { graphql } from '~/client/graphql';
+import { revalidate } from '~/client/revalidate-target';
+import { TAGS } from '~/client/tags';
+import { redirect } from '~/i18n/routing';
+import { serverToast } from '~/lib/server-toast';
+
+const VariantIdFromSkuQuery = graphql(`
+  query VariantIdFromSkuQuery($productId: Int!, $sku: String!) {
+    site {
+      product(entityId: $productId) {
+        variants(skus: [$sku]) {
+          edges {
+            node {
+              entityId
+            }
+          }
+        }
+      }
+    }
+  }
+`);
+
+const AddToWishlistMutation = graphql(`
+  mutation AddToWishlistMutation($wishlistId: Int!, $productId: Int!, $variantId: Int) {
+    wishlist {
+      addWishlistItems(
+        input: {
+          entityId: $wishlistId
+          items: [{ productEntityId: $productId, variantEntityId: $variantId }]
+        }
+      ) {
+        result {
+          entityId
+        }
+      }
+    }
+  }
+`);
+
+const DeleteWishlistItemsMutation = graphql(`
+  mutation DeleteWishlistItemsMutation($wishlistId: Int!, $wishlistItemId: Int!) {
+    wishlist {
+      deleteWishlistItems(input: { entityId: $wishlistId, itemEntityIds: [$wishlistItemId] }) {
+        result {
+          entityId
+        }
+      }
+    }
+  }
+`);
+
+const menuItemSchema = z.object({
+  action: z.enum(['add', 'remove']),
+  wishlistId: z.number().nonnegative(),
+  wishlistItemId: z.number().optional(),
+  redirectTo: z.string(),
+});
+
+const menuItemParser = z.string().transform((str, ctx) => {
+  try {
+    return menuItemSchema.parse(JSON.parse(str));
+  } catch {
+    ctx.addIssue({ code: 'custom', message: 'Invalid menuItem payload.' });
+
+    return z.NEVER;
+  }
+});
+
+const schema = z.object({
+  productId: z.number().nonnegative().min(1),
+  selectedSku: z.string(),
+  menuItem: menuItemParser,
+});
+
+interface WishlistAddMutationVariables {
+  wishlistId: number;
+  productId: number;
+  variantId?: number;
+}
+
+interface WishlistRemoveMutationVariables {
+  wishlistId: number;
+  wishlistItemId: number;
+}
+
+const getVariantIdFromSku = cache(
+  async (productId: number, sku: string, customerAccessToken?: string) => {
+    const { data } = await client.fetch({
+      document: VariantIdFromSkuQuery,
+      variables: { productId, sku },
+      customerAccessToken,
+      fetchOptions: { next: { revalidate } },
```

Comment:
Should we skip cache for logged in customer?

Reply:
Then we shouldn't pass the `customerAccessToken` no? I thought the CAT would be necessary for hidden variants. 

In terms of Vercel cache, we don't want to cache stuff that will be specific to a customer, since we don't want to spam the cache storage for each customer.

Reply:
Jorge is right here, we need to not cache this query if we are using a `customerAccessToken`. Next doesn't use headers as part of the cache key, so if Customer A has access to the product/variant and Customer B doesn't, whatever customer fetches this query first, the other would get the data from the other customer.

Reply:
Fair enough!